Need support for map options

Hi, I have a proto file that compiles fine with the protoc compiler, however, the wire compiler fails to compile it.
By the way, I couldn’t find anything on how to exclude a proto file from compiling, all the exclude keywords works after compilation.

here's the [build scan](https://scans.gradle.com/s/3c62sk5oble4w/console-log?anchor=194)

```
:generateMainProtos FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generateMainProtos'.
> class java.util.LinkedHashMap cannot be cast to class java.util.List (java.util.LinkedHashMap and java.util.List are in module java.base of loader 'bootstrap')

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':generateMainProtos'.
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$3(ExecuteActionsTaskExecuter.java:186)
	at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:268)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:184)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:173)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
	at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
	at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
	at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
	at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:411)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:398)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:391)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:377)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.ClassCastException: class java.util.LinkedHashMap cannot be cast to class java.util.List (java.util.LinkedHashMap and java.util.List are in module java.base of loader 'bootstrap')
	at com.squareup.wire.schema.Options.canonicalizeValue(Options.kt:195)
	at com.squareup.wire.schema.Options.canonicalizeValue(Options.kt:186)
	at com.squareup.wire.schema.Options.canonicalizeValue(Options.kt:186)
	at com.squareup.wire.schema.Options.canonicalizeOption(Options.kt:152)
	at com.squareup.wire.schema.Options.link(Options.kt:68)
	at com.squareup.wire.schema.EnumConstant.linkOptions$wire_schema(EnumConstant.kt:36)
	at com.squareup.wire.schema.EnumType.linkOptions(EnumType.kt:51)
	at com.squareup.wire.schema.MessageType.linkOptions(MessageType.kt:121)
	at com.squareup.wire.schema.FileLinker.linkOptions(FileLinker.kt:148)
	at com.squareup.wire.schema.Linker.link(Linker.kt:104)
	at com.squareup.wire.schema.SchemaLoader.loadSchema(SchemaLoader.kt:92)
	at com.squareup.wire.schema.WireRun.execute(WireRun.kt:222)
	at com.squareup.wire.schema.WireRun.execute(WireRun.kt:213)
	at com.squareup.wire.schema.WireRun.execute$default(WireRun.kt:211)
	at com.squareup.wire.gradle.WireTask.generateWireFiles(WireTask.kt:134)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:58)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:51)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:29)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$2.run(ExecuteActionsTaskExecuter.java:494)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:71)
	at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:71)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:479)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:462)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$400(ExecuteActionsTaskExecuter.java:105)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.executeWithPreviousOutputFiles(ExecuteActionsTaskExecuter.java:273)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:251)
	at org.gradle.internal.execution.steps.ExecuteStep.lambda$executeOperation$1(ExecuteStep.java:66)
	at org.gradle.internal.execution.steps.ExecuteStep.executeOperation(ExecuteStep.java:66)
	at org.gradle.internal.execution.steps.ExecuteStep.access$000(ExecuteStep.java:34)
	at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:47)
	at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:44)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
	at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
	at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:44)
	at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:34)
	at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:72)
	at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:42)
	at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:53)
	at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:39)
	at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:44)
	at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:77)
	at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:58)
	at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:54)
	at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:32)
	at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.execute(CaptureStateAfterExecutionStep.java:57)
	at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.execute(CaptureStateAfterExecutionStep.java:38)
	at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:63)
	at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:30)
	at org.gradle.internal.execution.steps.BuildCacheStep.executeWithoutCache(BuildCacheStep.java:176)
	at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:76)
	at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:47)
	at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:43)
	at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:32)
	at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:39)
	at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:25)
	at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:102)
	at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:95)
	at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:55)
	at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:39)
	at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:83)
	at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:44)
	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:37)
	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:27)
	at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:96)
	at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:52)
	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:83)
	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:54)
	at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:74)
	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.lambda$execute$2(SkipEmptyWorkStep.java:88)
	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:88)
	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:34)
	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:38)
	at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:46)
	at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:34)
	at org.gradle.internal.execution.steps.AssignWorkspaceStep.lambda$execute$0(AssignWorkspaceStep.java:43)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution$3.withWorkspace(ExecuteActionsTaskExecuter.java:286)
	at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:43)
	at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:33)
	at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:40)
	at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:30)
	at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:54)
	at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:40)
	at org.gradle.internal.execution.impl.DefaultExecutionEngine.execute(DefaultExecutionEngine.java:41)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:183)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:183)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:173)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
	at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
	at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
	at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
	at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:411)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:398)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:391)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:377)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)


* Get more help at https://help.gradle.org

BUILD FAILED in 1s
1 actionable task: 1 executed
```

--- a/wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Linker.kt
+++ b/wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Linker.kt
@@ -348,9 +348,9 @@ class Linker {
     requestedFields.add(field)
   }
 
-  /** Returns the field named [field] on the message type of [self]. */
+  /** Returns the field named [field] on the message type of [protoType]. */
   fun dereference(
-    self: Field,
+    protoType: ProtoType,
     field: String
   ): Field? {
     @Suppress("NAME_SHADOWING") var field = field
@@ -358,7 +358,7 @@ class Linker {
       field = field.substring(1, field.length - 1)
     }
 
-    val type = getForOptions(self.type!!)
+    val type = getForOptions(protoType)
     if (type is MessageType) {
       val messageField = type.field(field)
       if (messageField != null) return messageField

--- a/wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Options.kt
+++ b/wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Options.kt
@@ -101,7 +101,7 @@ class Options(
         // Retry with one upper level package to resolve relative paths.
         if (path == null) {
           packageName = packageName.substringBeforeLast(".", missingDelimiterValue = "")
-          if (packageName.isNullOrBlank() && !checkedExtensionFields) {
+          if (packageName.isBlank() && !checkedExtensionFields) {
             checkedExtensionFields = true
             val extensionFields = type.extensionFields.filter { it.name == option.name }
             if (extensionFields.size > 1) {
@@ -145,11 +145,13 @@ class Options(
       }
 
       last = nested
-      field = linker.dereference(field, path[i]) ?: return null // Unable to dereference segment.
+      field =
+        linker.dereference(field.type!!, path[i]) ?: return null // Unable to dereference segment.
       linker.request(field)
     }
 
-    last[get(lastProtoType!!, field!!)] = canonicalizeValue(linker, field, option.value)
+    last[get(lastProtoType!!, field!!)] =
+      canonicalizeValue(linker, field.type!!, field.isRepeated, option.value)
 
     check(result.size == 1) // TODO(benoit) might be safe to remove
     val (protoMember, value) = result.entries.first()
@@ -158,7 +160,8 @@ class Options(
 
   private fun canonicalizeValue(
     linker: Linker,
-    context: Field,
+    context: ProtoType,
+    isRepeated: Boolean,
     value: Any
   ): Any {
     when (value) {
@@ -166,43 +169,63 @@ class Options(
         val result = mutableMapOf<ProtoMember, Any>()
         val field = linker.dereference(context, value.name)
         if (field == null) {
-          linker.errors += "unable to resolve option ${value.name} on ${context.type}"
+          linker.errors += "unable to resolve option ${value.name} on $context"
         } else {
-          val protoMember = get(context.type!!, field)
-          result[protoMember] = canonicalizeValue(linker, field, value.value)
+          val protoMember = get(context, field)
+          result[protoMember] =
+            canonicalizeValue(linker, field.type!!, field.isRepeated, value.value)
         }
-        return coerceValueForField(context, result)
+        return coerceValueForField(context, result, isRepeated)
       }
 
       is Map<*, *> -> {
-        val result = mutableMapOf<ProtoMember, Any>()
-        for (entry in value) {
-          val name = entry.key as String
-          val field = linker.dereference(context, name)
-          if (field == null) {
-            linker.errors += "unable to resolve option $name on ${context.type}"
-          } else {
-            val protoMember = get(context.type!!, field)
-            result[protoMember] = canonicalizeValue(linker, field, entry.value!!)
+        if (context.isMap) {
+          // Map fields are defined with two optional entries: `key` and 'value'.
+          val mapFieldKeyAsString = value["key"]
+          val mapFieldValueAsString = value["value"]
+          val mapFieldKey =
+            if (mapFieldKeyAsString == null) null
+            else canonicalizeValue(
+              linker, context.keyType!!, isRepeated = false, mapFieldKeyAsString
+            )
+          val mapFieldValue =
+            if (mapFieldValueAsString == null) null
+            else canonicalizeValue(
+              linker, context.valueType!!, isRepeated = false, mapFieldValueAsString
+            )
+          return coerceValueForField(context, mapOf(mapFieldKey to mapFieldValue), isRepeated)
+
+        } else {
+          val result = mutableMapOf<ProtoMember, Any>()
+          for (entry in value) {
+            val name = entry.key as String
+            val field = linker.dereference(context, name)
+            if (field == null) {
+              linker.errors += "unable to resolve option $name on $context"
+            } else {
+              val protoMember = get(context, field)
+              result[protoMember] =
+                canonicalizeValue(linker, field.type!!, field.isRepeated, entry.value!!)
+            }
           }
+          return coerceValueForField(context, result, isRepeated)
         }
-        return coerceValueForField(context, result)
       }
 
       is List<*> -> {
         val result = mutableListOf<Any>()
         for (element in value) {
-          result.addAll(canonicalizeValue(linker, context, element!!) as List<Any>)
+          result.addAll(canonicalizeValue(linker, context, isRepeated, element!!) as List<Any>)
         }
-        return coerceValueForField(context, result)
+        return coerceValueForField(context, result, isRepeated)
       }
 
       is String -> {
-        return coerceValueForField(context, value)
+        return coerceValueForField(context, value, isRepeated)
       }
 
       is OptionElement.OptionPrimitive -> {
-        return canonicalizeValue(linker, context, value.value)
+        return canonicalizeValue(linker, context, isRepeated, value.value)
       }
 
       else -> {
@@ -211,9 +234,9 @@ class Options(
     }
   }
 
-  private fun coerceValueForField(context: Field, value: Any): Any {
+  private fun coerceValueForField(context: ProtoType, value: Any, isRepeated: Boolean): Any {
     return when {
-      context.isRepeated -> value as? List<*> ?: listOf(value)
+      isRepeated || context.isMap -> value as? List<*> ?: listOf(value)
       value is List<*> -> value.single()!!
       else -> value
     }

--- a/wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Schema.kt
+++ b/wire-library/wire-schema/src/commonMain/kotlin/com/squareup/wire/schema/Schema.kt
@@ -17,6 +17,7 @@ package com.squareup.wire.schema
 
 import com.squareup.wire.ProtoAdapter
 import kotlin.collections.set
+import okio.Path
 
 /**
  * A collection of .proto files that describe a set of messages. A schema is *linked*: each
@@ -41,6 +42,9 @@ class Schema internal constructor(protoFiles: Iterable<ProtoFile>) {
 
   /** Returns the proto file at [path], or null if this schema has no such file.  */
   fun protoFile(path: String): ProtoFile? = protoFiles.firstOrNull { it.location.path == path }
+
+  /** Returns the proto file at [path], or null if this schema has no such file.  */
+  fun protoFile(path: Path): ProtoFile? = protoFile(path.toString())
 
   /** Returns the proto file containing this [protoType], or null if there isn't such file.  */
   fun protoFile(protoType: ProtoType): ProtoFile? = protoFilesIndex[protoType]

--- a/wire-library/wire-schema/src/jvmTest/kotlin/com/squareup/wire/schema/OptionsTest.kt
+++ b/wire-library/wire-schema/src/jvmTest/kotlin/com/squareup/wire/schema/OptionsTest.kt
@@ -139,20 +139,20 @@ class OptionsTest {
 
     val protoFile = schema.protoFile("foo.proto")
 
-    val optionElements =
-      protoFile!!.types.stream().filter { it is MessageType && it.toElement().name == "Message" }
-        .findFirst().get().options.elements
+    val optionElements = protoFile!!.types
+      .first { it is MessageType && it.toElement().name == "Message" }
+      .options.elements
 
     assertThat(optionElements[0].toSchema())
       .isEqualTo(
         """|(foo) = {
-                      |  name: "test",
-                      |  type: STRING,
-                      |  schemes: [
-                      |    HTTP,
-                      |    HTTPS
-                      |  ]
-                      |}""".trimMargin()
+           |  name: "test",
+           |  type: STRING,
+           |  schemes: [
+           |    HTTP,
+           |    HTTPS
+           |  ]
+           |}""".trimMargin()
       )
 
     val foo = ProtoMember.get(Options.MESSAGE_OPTIONS, "foo")
@@ -273,5 +273,152 @@ class OptionsTest {
   @Test
   fun resolveFieldPathDoesntMatch() {
     assertThat(Options.resolveFieldPath("a.b", setOf("c", "d"))).isNull()
+  }
+
+  @Test
+  fun mapFieldEntriesLinking() {
+    val schema = buildSchema {
+      add("my_package/some_enum.proto".toPath(), """
+        |syntax = "proto2";
+        |import "google/protobuf/descriptor.proto";
+        |package my_package;
+        |
+        |enum SomeEnum {
+        |  TEST = 0 [(my_package.my_option) = {
+        |    entries: [
+        |      {
+        |        key: 'key-1',
+        |        value: {
+        |          some_string: "value-1"
+        |        }
+        |      },
+        |      {
+        |        value: {
+        |          some_string: "value-2"
+        |        }
+        |      },
+        |      {
+        |        key: 'key-3',
+        |        value: {
+        |        }
+        |      },
+        |      {
+        |        key: 'key-4',
+        |      },
+        |      {
+        |        key: 'key-5',
+        |        value: {
+        |          some_string: "value-5",
+        |          some_int32: 5
+        |        }
+        |      }
+        |    ]
+        |  }];
+        |}
+        |
+        |message SomeMessage {
+        |  optional string some_string = 1;
+        |  optional int32 some_int32 = 2;
+        |}
+        |
+        |message MyOption {
+        |  map<string, SomeMessage> entries = 2;
+        |}
+        |
+        |extend google.protobuf.EnumValueOptions {
+        |  optional MyOption my_option = 1000;
+        |}
+      """.trimMargin())
+    }
+
+    val enumType = schema.getType(ProtoType.get("my_package.SomeEnum")) as EnumType
+    val myOption = ProtoMember.get(Options.ENUM_VALUE_OPTIONS, "my_package.my_option")
+    val entries = ProtoMember.get(ProtoType.get("my_package.MyOption"), "entries")
+    val someString = ProtoMember.get(ProtoType.get("my_package.SomeMessage"), "some_string")
+    val someInt32 = ProtoMember.get(ProtoType.get("my_package.SomeMessage"), "some_int32")
+    assertThat(enumType.constant(tag = 0)!!.options.map)
+      .isEqualTo(
+        mapOf(
+          myOption to mapOf(
+            entries to listOf(
+              mapOf("key-1" to mapOf(someString to "value-1")),
+              mapOf(null to mapOf(someString to "value-2")),
+              mapOf("key-3" to mapOf()),
+              mapOf("key-4" to null),
+              mapOf("key-5" to mapOf(someString to "value-5", someInt32 to "5")),
+            )
+          )
+        )
+      )
+  }
+
+  @Test
+  fun mapFieldEntriesWriting() {
+    val path = "my_package/some_enum.proto".toPath()
+    val schema = buildSchema {
+      add(path, """
+        |syntax = "proto3";
+        |import "google/protobuf/descriptor.proto";
+        |package my_package;
+        |
+        |enum SomeEnum {
+        |  TEST = 0 [(my_package.my_option) = {
+        |    entries: [
+        |      {
+        |        key: 'key-1',
+        |        value: {
+        |          some_string: "value-1"
+        |        }
+        |      },
+        |      {
+        |        key: 'key-2',
+        |        value: {
+        |          some_string: "value-2",
+        |          some_int32: 2
+        |        }
+        |      }
+        |    ]
+        |  }];
+        |}
+        |
+        |message SomeMessage {
+        |  string some_string = 1;
+        |  int32 some_int32 = 2;
+        |}
+        |
+        |message MyOption {
+        |  map<string, SomeMessage> entries = 2;
+        |}
+        |
+        |extend google.protobuf.EnumValueOptions {
+        |  MyOption my_option = 1000;
+        |}
+      """.trimMargin())
+    }
+
+    val enumType = schema.getType(ProtoType.get("my_package.SomeEnum")) as EnumType
+    val optionElement=  enumType.constant(tag = 0)!!.options.elements.first()
+    // We do print "key" and "value" keys for map fields, even though the linked schema doesn't
+    // know about them.
+    val expected = """
+      |(my_package.my_option) = {
+      |  entries: [
+      |    {
+      |      key: "key-1",
+      |      value: {
+      |        some_string: "value-1"
+      |      }
+      |    },
+      |    {
+      |      key: "key-2",
+      |      value: {
+      |        some_string: "value-2",
+      |        some_int32: 2
+      |      }
+      |    }
+      |  ]
+      |}
+    """.trimMargin()
+    assertThat(optionElement.toSchema()).isEqualTo(expected)
   }
 }

